### PR TITLE
Add method to retreive values from several manifests

### PR DESCRIPTION
**Pull request by nesterov-n**
This reffers #31 
Add static method returning collections of values with specified key defined in several manifests

### CODE REVIEW

**Comment by dmarkov** [comment]
@nesterov-n Many thanks for the PR, let me find a reviewer for it

**Comment by dmarkov** [comment]
@mkordas please review, thanks

**Comment by mkordas** [comment]
@nesterov-n I'm on it

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 1]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
...
+    /**
+     * Reads all values with specified key from several manifests.
+     * @param manifests Manifests
+     * @param key Value key
+     * @return Collections of values.
+     * @throws IOException If some I/O problem inside
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public static Collection<String> all(final Mfs manifests,
+                                         final String key) throws IOException {
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n we do not align parameters like that, let's move them to the left

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 2]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
         this.attributes = new ConcurrentHashMap<String, String>(attrs);
     }
 
+    /**
+     * Reads all values with specified key from several manifests.
+     * @param manifests Manifests
+     * @param key Value key
+     * @return Collections of values.
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n dot here is unnecessary

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 3]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
         this.attributes = new ConcurrentHashMap<String, String>(attrs);
     }
 
+    /**
+     * Reads all values with specified key from several manifests.
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n I think it should be more descriptive. What if there is no such key in any manifest or there is same key with same/different values across manifests?

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 4]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
...
+     * @param manifests Manifests
+     * @param key Value key
+     * @return Collections of values.
+     * @throws IOException If some I/O problem inside
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public static Collection<String> all(final Mfs manifests,
+                                         final String key) throws IOException {
+        final Collection<InputStream> streams = manifests.fetch();
+        final Collection<String> values = new LinkedList<String>();
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n see http://stackoverflow.com/a/323889/4647853, let's use array list

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 5]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
...
+     * Reads all values with specified key from several manifests.
+     * @param manifests Manifests
+     * @param key Value key
+     * @return Collections of values.
+     * @throws IOException If some I/O problem inside
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public static Collection<String> all(final Mfs manifests,
+                                         final String key) throws IOException {
+        final Collection<InputStream> streams = manifests.fetch();
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n can we inline this variable?

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 6]
```diff
@@ -162,6 +163,29 @@ public Manifests(final Map<String, String> attrs) {
...
+        final Collection<String> values = new LinkedList<String>();
+        for (final InputStream stream : streams) {
+            final String value = new Manifests().append(
+                    new StreamsMfs(stream)
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n we do not indent more than 4 spaces compared to previous line

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 7]
```diff
@@ -42,6 +46,7 @@
  * @author Yegor Bugayenko (yegor@tpc2.com)
  * @version $Id$
  * @since 0.7
+ * @checkstyle MultipleStringLiteralsCheck (120 lines)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n can we somehow avoid this suppression?

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 8]
```diff
@@ -127,4 +132,34 @@ public void appendsAttributesFromInputStream() throws Exception {
         );
     }
 
+    /**
+     * Manifests can read values from several manifests.
+     * @throws Exception If something goes wrong
+     * @since 2.0
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n I think it should be just 1.2

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 9]
```diff
@@ -127,4 +132,34 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+        MatcherAssert.assertThat(
+                values.contains("some test attribute"),
+                Matchers.equalTo(true)
+        );
+        MatcherAssert.assertThat(
+                values.contains("another test attribute"),
+                Matchers.equalTo(true)
+        );
+    }
+
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n we need more tests here: 0 manifests, 2 manifests but only one with the key etc.

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 10]
```diff
@@ -127,4 +132,34 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+        MatcherAssert.assertThat(
+                values.size(),
+                Matchers.equalTo(2)
+        );
+        MatcherAssert.assertThat(
+                values.contains("some test attribute"),
+                Matchers.equalTo(true)
+        );
+        MatcherAssert.assertThat(
+                values.contains("another test attribute"),
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n can we use contains matcher instead?

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 11]
```diff
@@ -127,4 +132,34 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+                this.getClass().getResourceAsStream("another-test.mf")
+        );
+        final Mfs streamManifests = new StreamsMfs(streams);
+        final Collection<String> values = Manifests.all(
+                streamManifests,
+                "From-File"
+        );
+        MatcherAssert.assertThat(
+                values.size(),
+                Matchers.equalTo(2)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n there is matcher for size of collection

**Comment by mkordas** [comment]
@nesterov-n see my review above

**Comment by mkordas** [comment]
@nesterov-n ping

**Comment by mkordas** [comment]
@nesterov-n do you need any help?

**Comment by nesterov-n** [comment]
@mkordas Sorry for delay.  I fixed your hints, please check

**Comment by mkordas** [comment]
@nesterov-n I'll check it now

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 12]
```diff
@@ -162,6 +163,31 @@ public Manifests(final Map<String, String> attrs) {
...
+     * such key in any manifests.
+     * @param manifests Manifests
+     * @param key Value key
+     * @return Collections of values
+     * @throws IOException If some I/O problem inside
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public static Collection<String> all(final Mfs manifests, final String key)
+        throws IOException {
+        final Collection<String> values = new ArrayList<String>(0);
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n what's the point in having list with 0 initial capacity? We should have at least a few places to begin with

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 13]
```diff
@@ -44,6 +49,14 @@
  * @since 0.7
  */
 public final class ManifestsTest {
+    // @checkstyle JavadocVariableCheck (8 lines)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n have you seen that practice in other files? Generally we have Javadocs everywhere

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 14]
```diff
@@ -44,6 +49,14 @@
...
+    public static final String TEST_MANIFEST = "test.mf";
+    public static final String TEST_MANIFEST_2 = "another-test.mf";
+    public static final String TEST_ATTRIBUTE = "some test attribute";
+    public static final String TEST_ATTRIBUTE_2 = "another attribute";
+    public static final String FROM_FILE = "From-File";
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n can we have these fields `private`?

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 15]
```diff
@@ -44,6 +49,14 @@
...
+    public static final String VERSION = "Version";
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n please see that article, we should not have such constants: http://www.yegor256.com/2015/07/06/public-static-literals.html

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 16]
```diff
@@ -44,6 +49,14 @@
  * @since 0.7
  */
 public final class ManifestsTest {
+    // @checkstyle JavadocVariableCheck (8 lines)
+    public static final String TEST_MANIFEST = "test.mf";
+    public static final String TEST_MANIFEST_2 = "another-test.mf";
+    public static final String TEST_ATTRIBUTE = "some test attribute";
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n 
> Another typical mistake is to use constants in unit tests to avoid duplicate string/numeric literals in test methods. Don't do this! Every test method should work with its own set of input values.

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 17]
```diff
@@ -119,11 +132,103 @@ public void appendsAttributesFromFile() throws Exception {
...
+        );
+        MatcherAssert.assertThat(
+            values,
+            Matchers.hasSize(2)
+        );
+        MatcherAssert.assertThat(
+            values,
+            Matchers.contains(
+                TEST_ATTRIBUTE,
+                    TEST_ATTRIBUTE_2
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n bit strange formatting, it should be on the same level as item above

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 18]
```diff
@@ -119,11 +132,103 @@ public void appendsAttributesFromFile() throws Exception {
...
+            )
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n but... doesn't it fit one line?

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 19]
```diff
@@ -119,11 +132,103 @@ public void appendsAttributesFromFile() throws Exception {
...
+            this.getClass().getResourceAsStream(TEST_MANIFEST),
+            this.getClass().getResourceAsStream(TEST_MANIFEST_2)
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            FROM_FILE
+        );
+        MatcherAssert.assertThat(
+            values,
+            Matchers.hasSize(2)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n this fits one line, let's join them

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 20]
```diff
@@ -119,11 +132,103 @@ public void appendsAttributesFromFile() throws Exception {
...
+            this.getClass().getResourceAsStream(TEST_MANIFEST),
+            this.getClass().getResourceAsStream(TEST_MANIFEST_2)
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            MISSED_KEY
+        );
+        MatcherAssert.assertThat(
+            values,
+            Matchers.hasSize(0)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n let's try putting as much on every line as possible

File: src/main/java/com/jcabi/manifests/Manifests.java [hunk 21]
```diff
@@ -162,6 +163,31 @@ public Manifests(final Map<String, String> attrs) {
         this.attributes = new ConcurrentHashMap<String, String>(attrs);
     }
 
+    /**
+     * Reads all values with specified key from several manifests.
+     * Result collections contains all values with specified key
```
Comment (mkordas) [inline src/main/java/com/jcabi/manifests/Manifests.java]: @nesterov-n `collections`->`collection`

**Comment by mkordas** [comment]
@nesterov-n see my comments above

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 22]
```diff
@@ -119,11 +132,103 @@ public void appendsAttributesFromFile() throws Exception {
...
+            this.getClass().getResourceAsStream(TEST_MANIFEST),
+            this.getClass().getResourceAsStream(TEST_MANIFEST_2)
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            VERSION
+        );
+        MatcherAssert.assertThat(
+            values,
+            Matchers.hasSize(1)
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n please read that: http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#indentation
> you should put as much as possible on one line - within the limit of 80 characters

**Comment by nesterov-n** [comment]
@mkordas Thanks for hints. Fixed

**Comment by mkordas** [comment]
@nesterov-n I'm checking

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 23]
```diff
@@ -127,4 +132,72 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+    public void readsValuesExistingInSomeManifests() throws Exception {
+        final List<InputStream> streams = Arrays.asList(
+            this.getClass().getResourceAsStream("test4.mf"),
+            this.getClass().getResourceAsStream("test5.mf")
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            "Version"
+        );
+        MatcherAssert.assertThat(values, Matchers.hasSize(1));
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n this assertion is covered below

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 24]
```diff
@@ -127,4 +132,72 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+    public void returnsEmptyCollectionForMissedKey() throws Exception {
+        final List<InputStream> streams = Arrays.asList(
+            this.getClass().getResourceAsStream("test6.mf"),
+            this.getClass().getResourceAsStream("test7.mf")
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            "Missed-key"
+        );
+        MatcherAssert.assertThat(values, Matchers.hasSize(0));
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n this could be just empty matcher

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 25]
```diff
@@ -127,4 +132,72 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+    public void readsValuesFromSeveralManifests() throws Exception {
+        final List<InputStream> streams = Arrays.asList(
+            this.getClass().getResourceAsStream("test2.mf"),
+            this.getClass().getResourceAsStream("test3.mf")
+        );
+        final Collection<String> values = Manifests.all(
+            new StreamsMfs(streams),
+            "Name"
+        );
+        MatcherAssert.assertThat(values, Matchers.hasSize(2));
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n this assertion is obsolete, it is covered below

File: src/test/java/com/jcabi/manifests/ManifestsTest.java [hunk 26]
```diff
@@ -127,4 +132,72 @@ public void appendsAttributesFromInputStream() throws Exception {
...
+    /**
+     * Manifests can return empty collection for empty manifest list.
+     * @throws Exception If something goes wrong
+     * @since 1.2
+     */
+    @Test
+    public void returnsEmptyCollectionForEmptyManifestList() throws Exception {
+        final Mfs empty = new StreamsMfs(Collections.<InputStream>emptyList());
+        final Collection<String> values = Manifests.all(empty, "Key");
+        MatcherAssert.assertThat(values, Matchers.hasSize(0));
```
Comment (mkordas) [inline src/test/java/com/jcabi/manifests/ManifestsTest.java]: @nesterov-n same here, just empty

**Comment by mkordas** [comment]
@nesterov-n just comments about matchers, all else is good

**Comment by nesterov-n** [comment]
@mkordas Done.

P.S Is there some instrument to check code style: line breaks, indentation, etc. Seems that qulice doesn't cover all cases. Maybe someone has xml with code style for IDEA or some other tool? Don't want to spend so many your time for further pull requests.

**Comment by mkordas** [comment]
@nesterov-n there are still two comments open, there is no need to assert size when you assert using contains

**Comment by mkordas** [comment]
@nesterov-n this project just uses old Qulice, but even in latest version not all features are present already. IDEA settings are here: http://img.teamed.io/settings.jar

**Comment by nesterov-n** [comment]
@mkordas Sorry, missed some of your comments. Removed redundant size matchers from tests. Many thanks for IDEA settings.

**Comment by mkordas** [comment]
@nesterov-n thanks, looks good!

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@nesterov-n I don't really like this... instead of doing this, we should change the way we **store** data. we need to store them in `Map<String, List<String>>`

**Comment by nesterov-n** [comment]
@yegor256 Can you close this PR. I will make new one

**Comment by mkordas** [comment]
@nesterov-n generally the rule is to have just one PR (http://at.teamed.io/qa.html), so I'm not sure if @yegor256 is OK to continue in another one

**Comment by nesterov-n** [comment]
@mkordas No problem I will update this PR when issue #31 is assigned to me again

**Comment by mkordas** [comment]
@nesterov-n I think we need to close this one. There's limit of 2 tasks in this project, but one slot is taken by this PR...

**Comment by nesterov-n** [comment]
@mkordas no problem

**Comment by dmarkov** [comment]
@mkordas Thanks so much! Your account was topped up for **55 mins** (transaction ID is `5715fd2e00d77d2ec2000502`, the task took 260 hours and 18 mins); you're getting extra minutes here (c=40); +55 to your rating, your total score is [+8302](http://www.netbout.com/b/36080?open=rating)
